sequencer/src/block2.rs

```diff
@@ -58,13 +66,168 @@ impl BlockPayload {
         payload.extend(tx_bodies);
         Self { payload }
     }
+
+    // Return the range `r` such that the `index`th tx bytes are at `self.payload[r]`.
+    fn get_tx_range(&self, index: TxIndex) -> Option<Range<usize>> {
+        let tx_bodies_offset = self.tx_bodies_offset()?;
+
+        // tx_table[i] is end index for the ith tx,
+        // so the range for the ith tx is tx_table[i-1..i].
+        // But tx_table starts at index 1 in the payload
+        // because index 0 in the payload is the length of the tx_table.
+        // So the range for the ith tx in the payload is actually
+        // [index..index+1].
+        let start = if index == 0 {
+            0
+        } else {
+            self.get_value_usize(index)?
+        }
+        .checked_add(tx_bodies_offset)?;
+
+        let end = self
+            .get_value_usize(index.checked_add(1)?)?
+            .checked_add(tx_bodies_offset)?;
+
+        Some(start..end)
+    }
+
+    // Viewing `self.payload` bytes as a vec of words of type `TxIndex`,
+    // return the `index`th word.
+    fn get_value(&self, index: TxIndex) -> Option<TxIndex> {
+        // TODO idiomatic usize <-> TxIndex conversion.
+        let start = usize::try_from(index)
```

Comment:
IMO, we could treat it as an invariant that all the indices are small enough to fit in a `u32`, and panic if any of these conversions fail. We just need to check and return an error whenever we are building a block or receiving a block we don't necessarily know is well-formed (e.g. `from_transactions` and `deserialize`). Then we could simplify error handling a lot everywhere else.

Reply:
I introduced a newtype `TxTableEntry` for type safety and clarity as per https://github.com/EspressoSystems/espresso-sequencer/pull/776#discussion_r1405992221 . Now it's equally crufty to panic vs. check errors. Also, all arithmetic is paranoid with `checked_add` etc. Now someone would need to try pretty hard to accidentally introduce a bug.

(Currently my "errors" are only `Option` but this can be changed later to `Result` if desired. Frankly, I'm getting sick of fighting with `Result`.)